### PR TITLE
Motoko Intro slides

### PR DESCRIPTION
**Pull request by crusso**
These are the slides I prepared for the Hackathon intro to Motoko.

Done as adoc, so we can eventually use them online as part of main documentation.

Dfinity internal presentation here (just screen shot to Google slides)

https://docs.google.com/presentation/d/1uW4JBSXfJwtrtG8W5aivSmQp3d_D-cfwrRX7htQfdHo/edit?usp=sharing

I'd like to check them in, but we can choose to omit them from the navigation if it needs more words...

### CODE REVIEW

File: doc/modules/language-guide/pages/overview.adoc [hunk 1]
```diff
@@ -0,0 +1,997 @@
+== Overview
+
+=== Motivation and Goals
+
+A simple, useful language for the Internet Computer (IC)
+
+* Familiar syntax
+* Safe by default
+* Incorporating [line-through]*actor* _canister_ model
```
Comment (rossberg) [inline doc/modules/language-guide/pages/overview.adoc:9]:
```suggestion
* Incorporating *actor* model for canisters
```

File: doc/modules/language-guide/pages/overview.adoc [hunk 2]
```diff
@@ -0,0 +1,997 @@
...
+* Seamless integration of IC features
+* Making most of present and future WebAssembly
+
+=== Key Design Points
+
+* Object-oriented, functional & imperative
+* Objects as records of members
```
Comment (rossberg) [inline doc/modules/language-guide/pages/overview.adoc:16]:
```suggestion
* Objects as records of functions
```

File: doc/modules/language-guide/pages/overview.adoc [hunk 3]
```diff
@@ -0,0 +1,997 @@
...
+* `async`/`await` for sequential programming of asynchronous messaging
+* Structural typing with simple generics and subtyping
+* Safe arithmetic (unbounded and checked)
```
Comment (rossberg) [inline doc/modules/language-guide/pages/overview.adoc:19]:
```suggestion
* Safe arithmetic (both unbounded and checked)
```

File: doc/modules/language-guide/pages/overview.adoc [hunk 4]
```diff
@@ -0,0 +1,997 @@
...
+* Non-nullable types by default
+* Garbage collected (no manual memory management)
+* JavaScript-like syntax but statically typed & sane
+
+Inspirations: Java(Script), C#, Swift, Pony, ML, Haskell
```
Comment (rossberg) [inline doc/modules/language-guide/pages/overview.adoc:24]:
```suggestion
Inspirations: Java, JavaScript, C#, Swift, Pony, ML, Haskell
```

File: doc/modules/language-guide/pages/overview.adoc [hunk 5]
```diff
@@ -0,0 +1,997 @@
...
+
+=== Semantics
+
+* call-by-value (like Java, C, JS, ML; unlike Haskell, Nix)
```
Comment (rossberg) [inline doc/modules/language-guide/pages/overview.adoc:28]:
```suggestion
* call-by-value (like Java, C, JS, ML; unlike Haskell)
```

File: doc/modules/language-guide/pages/overview.adoc [hunk 6]
```diff
@@ -0,0 +1,997 @@
...
+* declarations are locally mutually recursive
+* parametric, bounded polymorphism
+* subtyping as subsumption, not coercion.
```
Comment (rossberg) [inline doc/modules/language-guide/pages/overview.adoc:31]:
```suggestion
* subtyping as zero-cost subsumption, not coercion
```

File: doc/modules/language-guide/pages/overview.adoc [hunk 7]
```diff
@@ -0,0 +1,997 @@
...
+* no inheritance
+
+=== Implementation(s)
+
+* implemented in OCaml (leverages `wasm` libary)
+* simple reference interpreter
+* less simple compiler to WebAssembly
+** multipass with typed IR in each pass.
+** uniform representation, unboxed arithmetic
+** two-space gc or mark-compact gc, gc between messages (for now)
```
Comment (rossberg) [inline doc/modules/language-guide/pages/overview.adoc:42]:
```suggestion
** two-space GC or mark-compact GC, invoked after messages (for now)
```

File: doc/modules/language-guide/pages/overview.adoc [hunk 8]
```diff
@@ -0,0 +1,997 @@
...
+  import Trie "mo:base/Trie";
+
+  type Users = Trie.Trie<Text, Nat>; // reference types
+
+  Debug.print(Int.toText(7)); // reference functions/values
+....
+
+=== Blocks and declarations
+
+
```
Comment (rossberg) [inline doc/modules/language-guide/pages/overview.adoc:83]:
```suggestion
```

File: doc/modules/language-guide/pages/overview.adoc [hunk 9]
```diff
@@ -0,0 +1,997 @@
...
+* `if (b) … else …`
+* `switch e { case (pat1) e1; …; case _ en }`
+* `while (b) …`
+* `loop …`
+* `loop …  while (b)`
+* `for (pat in e) …`
+* `return`, `return e`
+* `label l e`,  `break l e`
+* `async e`, `await e`             _(restricted)_
+* `throw`, `try … catch (x) { … }` _(restricted)_
```
Comment (rossberg) [inline doc/modules/language-guide/pages/overview.adoc:115]:
```suggestion
* `if b …`
* `if b … else …`
* `switch e { case pat1 e1; …; case _ en }`
* `while b …`
* `loop …`
* `loop …  while b`
* `for (pat in e) …`
* `return`, `return e`
* `label l e`,  `break l e`
* `async e`, `await e`             _(restricted)_
* `throw`, `try … catch x { … }` _(restricted)_
```

File: doc/modules/language-guide/pages/overview.adoc [hunk 10]
```diff
@@ -0,0 +1,997 @@
...
+
```
Comment (rossberg) [inline doc/modules/language-guide/pages/overview.adoc:116]: Mention `do?` here?

File: doc/modules/language-guide/pages/overview.adoc [hunk 11]
```diff
@@ -0,0 +1,997 @@
...
+`Nat` is a _subtype_ of `Int`
+
+(you can supply a `Nat` wherever an `Int` is expected)
+
+=== Bounded numbers (trapping)
+
+`Nat8`, `Nat16`, `Nat32`, `Nat64`, `Int8`, `Int16`, `Int32`, `Int64`
+
+Trap on over- and underflow. Wrap-on-trap and bit-manipulating
+operations available.
```
Comment (rossberg) [inline doc/modules/language-guide/pages/overview.adoc:152]:
```suggestion
Trap on over- and underflow; wrap-around and bit-manipulating
operations available separately
```

File: doc/modules/language-guide/pages/overview.adoc [hunk 12]
```diff
@@ -0,0 +1,997 @@
...
+`a << b` +
+…
+
+`a +% b, a -% b, …` for wrapping, modular arithmetic (where appropriate)
+
+=== Characters and text
+
+`Char`, `Text`
+
+Unicode! No random access.
```
Comment (rossberg) [inline doc/modules/language-guide/pages/overview.adoc:185]:
```suggestion
Unicode! Character = Unicode scalar value; no random access on text
```

File: doc/modules/language-guide/pages/overview.adoc [hunk 13]
```diff
@@ -0,0 +1,997 @@
...
+`- x` +
+`a + b` +
+`a % b` +
+`a & b` +
+`a << b` +
+…
+
+`a +% b, a -% b, …` for wrapping, modular arithmetic (where appropriate)
+
+=== Characters and text
```
Comment (rossberg) [inline doc/modules/language-guide/pages/overview.adoc:181]:
```suggestion
=== Characters and Text
```

File: doc/modules/language-guide/pages/overview.adoc [hunk 14]
```diff
@@ -0,0 +1,997 @@
...
+[source.include_display, motoko]
+....
+display(null);
+....
+
+[source.include_display, motoko]
+....
+display(?"Test");
+....
+
```
Comment (rossberg) [inline doc/modules/language-guide/pages/overview.adoc:302]: Simple `do?` example here?

File: doc/modules/language-guide/pages/overview.adoc [hunk 15]
```diff
@@ -0,0 +1,997 @@
...
+counters[1] := counters[1] + 1;
+
+// counters[3] will trap (fixed size)
+
+counters;
+....
+
+=== Records
+
+`{ first : Text; last : Text; salary : var Nat }`
```
Comment (rossberg) [inline doc/modules/language-guide/pages/overview.adoc:339]:
```suggestion
`{first : Text; last : Text; salary : var Nat}`
```

File: doc/modules/language-guide/pages/overview.adoc [hunk 16]
```diff
@@ -0,0 +1,997 @@
...
+
+[source.include_impDebugInt, motoko]
+....
+let employee = { first = "Boris"; last = "Johnson"; var salary = 81_932 };
```
Comment (rossberg) [inline doc/modules/language-guide/pages/overview.adoc:343]:
```suggestion
let employee = {first = "Boris"; last = "Johnson"; var salary = 81_932};
```

File: doc/modules/language-guide/pages/overview.adoc [hunk 17]
```diff
@@ -0,0 +1,997 @@
...
+
+Debug.print(
+  employee.first # " " # employee.last # " earns " #
+  Int.toText(employee.salary) # " pounds."
```
Comment (rossberg) [inline doc/modules/language-guide/pages/overview.adoc:347]:
```suggestion
    Int.toText(employee.salary) # " pounds."
```

File: doc/modules/language-guide/pages/overview.adoc [hunk 18]
```diff
@@ -0,0 +1,997 @@
...
+
+=== Objects
+
+`{ first : Text; last : Text; get : () -> Nat; add : Nat -> () }`
+
+[source, motoko]
+....
+object self {
+  public let first = "Boris";
+  public let last = "Johnson";
```
Comment (rossberg) [inline doc/modules/language-guide/pages/overview.adoc:363]: Seriously?

Comment (rossberg) [inline doc/modules/language-guide/pages/overview.adoc:363]: Seriously?

File: doc/modules/language-guide/pages/overview.adoc [hunk 19]
```diff
@@ -0,0 +1,997 @@
...
+  var salary : Nat = 81_932; // private by default
+  public func get() : Nat = salary;
+  public func add(bump : Nat) { salary += bump };
+}
+....
+
+=== Classes
+
+
```
Comment (rossberg) [inline doc/modules/language-guide/pages/overview.adoc:372]:
```suggestion
```

File: doc/modules/language-guide/pages/overview.adoc [hunk 20]
```diff
@@ -0,0 +1,997 @@
...
+  };
+};
+....
+
+=== Recursive Types
+
+[source#Lists, motoko]
+....
+type List = {
+  #item : { head : Text; tail : List }; // variant with payload!
```
Comment (rossberg) [inline doc/modules/language-guide/pages/overview.adoc:429]: Using a record here strikes me as rather verbose in use. I'd use:
```suggestion
  #cons : (head : Text, tail : List); // variant with payload!
```

File: doc/modules/language-guide/pages/overview.adoc [hunk 21]
```diff
@@ -0,0 +1,997 @@
...
+`base` package provides basic features as separate modules.
+
+More libraries popping up!
+
+`MyLib.mo` _must_ contain a module or actor class, eg:
+
+[source.no-repl, motoko]
+....
+module {
+
```
Comment (rossberg) [inline doc/modules/language-guide/pages/overview.adoc:512]:
```suggestion
```

File: doc/modules/language-guide/pages/overview.adoc [hunk 22]
```diff
@@ -0,0 +1,997 @@
...
+};
+....
+
+_sharable_ arguments and _no_ or _async_ result type.
+
+* `register` is a _oneway_ IC method (unawaitable).
+* `send` is an IC _update_ method
+* `recv` is IC _query_ method
+
+IC "canister" ≈ Motoko "actor"
```
Comment (rossberg) [inline doc/modules/language-guide/pages/overview.adoc:543]:
```suggestion
IC canister ≈ Motoko actor
```

File: doc/modules/language-guide/pages/overview.adoc [hunk 23]
```diff
@@ -0,0 +1,997 @@
...
+* objects (with methods)
+
+=== A complete actor
+
+[source, motoko]
+....
+import Array "mo:base/Array";
+
+actor Broadcast {
+
```
Comment (rossberg) [inline doc/modules/language-guide/pages/overview.adoc:568]:
```suggestion
```

**Review by rossberg — APPROVED** [review]

**Comment by crusso** [comment]
@nomeata any idea why this if failing CI?

**Pull request merged**